File: face_detection/face_detection.py
# Ref: https://youtu.be/mPCZLOVTEc4

import cv2
import numpy as np

face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
eye_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_eye.xml')
q
cap = cv2.VideoCapture(0)

while True:
    _, frame = cap.read()
    
    faces = face_cascade.detectMultiScale(frame)
    for x,y,w,h in faces:
        cv2.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
        eyes = eye_cascade.detectMultiScale(frame[y:y+h,x:x+w])
        for xe, ye, we, he in eyes:
            cv2.rectangle(frame,(x+xe,y+ye),(x+xe+we,y+ye+he),(0,0,255),2)
    cv2.imshow("Frame", frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()

File: cpp/face_detection/FaceDetection.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;

int main()
{
    CascadeClassifier face_cascade, eye_cascade;
    face_cascade.load("/usr/share/opencv4/haarcascades/haarcascade_frontalface_default.xml");
    eye_cascade.load("/usr/share/opencv4/haarcascades/haarcascade_eye.xml");
    VideoCapture cap(0);
    while (1)
    {
        Mat frame;
        cap.read(frame);
        std::vector<Rect> rects;
        face_cascade.detectMultiScale(frame, rects);
        for (Rect bbox : rects)
        {
            rectangle(frame, bbox, {0, 0, 255}, 2);
            std::vector<Rect> eyes;
            eye_cascade.detectMultiScale(frame.rowRange(bbox.y, bbox.y + bbox.height).colRange(bbox.x, bbox.x + bbox.width), eyes);

            for (Rect bbox_eye : eyes)
            {
                bbox_eye = Rect2i(bbox.x + bbox_eye.x, bbox.y + bbox_eye.y, bbox_eye.width, bbox_eye.height);
                rectangle(frame, bbox_eye, {255, 0, 0}, 2);
            }
        }

        imshow("", frame);
        char key = waitKey(1);
        if (key == 'q')
        {
            break;
        }
    }
    cap.release();
    destroyAllWindows();
    return 0;
}
